tools/golang/pkg/encryption/encryption.go

```diff
@@ -0,0 +1,295 @@
+/* Copyright 2019 The Subscribe with Google Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package encryption
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"github.com/golang/protobuf/proto"
+	"github.com/google/tink/go/aead"
+	"github.com/google/tink/go/core/registry"
+	"github.com/google/tink/go/hybrid"
+	"github.com/google/tink/go/insecurecleartextkeyset"
+	"github.com/google/tink/go/keyset"
+	gcmpb "github.com/google/tink/proto/aes_gcm_go_proto"
+	tinkpb "github.com/google/tink/proto/tink_go_proto"
+	"golang.org/x/net/html"
+	"golang.org/x/net/html/atom"
+	"net/http"
+	"strings"
+	"unicode/utf8"
+)
+
+// Helper functions for the SwG Encryption Script.
+
+const aesGcmKeyUrl string = "type.googleapis.com/google.crypto.tink.AesGcmKey"
+const aesGcmKeySize uint32 = 16
+
+// Public function to generate an encrypted HTML document given the original.
+func GenerateEncryptedDocument(htmlStr, accessRequirement string, pubKeys map[string]tinkpb.Keyset) (string, error) {
+	km, err := registry.GetKeyManager(aesGcmKeyUrl)
+	if err != nil {
+		return "", err
+	}
+	key, err := generateNewAesGcmKey(km)
+	if err != nil {
+		return "", err
+	}
+	ks := createAesGcmKeyset(key)
+	ksEnc, err := proto.Marshal(&ks)
+	if err != nil {
+		return "", err
+	}
+	kh, err := insecurecleartextkeyset.Read(&keyset.MemReaderWriter{Keyset: &ks})
+	if err != nil {
+		return "", err
+	}
+	parsedHtml, err := html.Parse(strings.NewReader(htmlStr))
+	if err != nil {
+		return "", err
+	}
+	encryptedSections := getAllEncryptedSections(parsedHtml)
+	if len(encryptedSections) == 0 {
+		return "", errors.New("No encrypted sections found.")
```

Comment:
Should this error or simply no-op?

Reply:
I like error out because the publisher is aware that nothing happened to the document instead of silently doing nothing or returning an empty string.